src/kubernetes/reflector.rs

```diff
@@ -0,0 +1,569 @@
+                        warn!(message = "got unexpected data in the watch response");
+                        continue;
+                    }
+                };
+
+                // Prepare a resource version cvandidate so we can update (aka
+                // commit) it later.
+                let resource_version_candidate =
+                    match resource_version::Candidate::from_watch_event(&event) {
+                        Some(val) => val,
+                        None => {
+                            // This event doesn't have a resource version, this means
+                            // it's not something we care about.
+                            continue;
+                        }
+                    };
+
+                // Process the event.
+                self.process_event(event);
+
+                // Record the resourse version for this event, so when we resume
+                // it won't be redelivered.
+                self.resource_version.update(resource_version_candidate);
+
+                // Flush the changes to the state.
+                self.state_writer.flush();
+            }
+
+            // For the next pause duration we won't get any updates.
+            // This is better than flooding k8s api server with requests.
+            delay_for(self.pause_between_requests).await;
+        }
+    }
+
+    /// Prepare and execute a watch request.
+    async fn issue_request(
+        &mut self,
+    ) -> Result<<W as Watcher>::Stream, watcher::invocation::Error<<W as Watcher>::InvocationError>>
+    {
+        let watch_optional = WatchOptional {
+            field_selector: self.field_selector.as_ref().map(|s| s.as_str()),
+            label_selector: self.label_selector.as_ref().map(|s| s.as_str()),
+            pretty: None,
+            resource_version: self.resource_version.get(),
+            timeout_seconds: None,
+            allow_watch_bookmarks: Some(true),
+        };
+        let stream = self.watcher.watch(watch_optional).await?;
+        Ok(stream)
+    }
+
+    /// Translate received watch event to the state update.
+    fn process_event(&mut self, event: WatchEvent<<W as Watcher>::Object>) {
+        match event {
+            WatchEvent::Added(object) => {
+                if let Some((key, value)) = kv(object) {
+                    trace!(message = "got an object event", uid = ?key, event = "added");
+                    self.state_writer.insert(key, value);
+                } else {
+                    warn!(
+                        message = "got an object event but unable to obtain a key from object",
+                        event = "added",
+                    );
+                }
+            }
+            WatchEvent::Deleted(object) => {
+                if let Some((key, _value)) = kv(object) {
+                    trace!(message = "got an object event", uid = ?key, event = "deleted");
+                    self.state_writer.empty(key);
```

Comment:
Right, this is one of the race conditions that we can't currently solve properly due to the limitations of the APIs.
Until we improve the APIs, we can only add mitigations to reduce the effect of it, but no proper fix. I saw the timeouts you had at the `WatchClient`, and I think I can add a similar approach here.
Though, I wonder if this is a real issue? I mean, would it make a difference if we add the timeout-based mitigation, or does it currently work mostly alright in practice? If you have conducted experiments before - please share them!

I'll note to add an integration test to check this race condition. This way if we'll be encountering it in CI a lot I'll apply the mitigation.

Reply:
Yes it happened a few times since there is a quite common scenario to trigger it:
  1. Have a running container with larger amount of logs in the log file.
  2. Start vector.
  3. Container can write some more logs here but it ends while the file server is still going through the log file.

Even if we discard all of the logs that happened before 2. we would still not enrich logs from 3. 
Now, as you mentioned, this is quite a race, and yes there is no real way to solve it at the moment, but I would consider the few last logs from a container of high importance since if the container ended in an error, it's highly likely that those error logs will be among the last logs, so it would be a shame if they aren't enriched.

Reply:
I agree! Thanks for providing a reproduction scenario, with this it will be easy to create a test case!

The mitigation would be to add a caching layer at the annotator and path provider. The cache would first look at the `evmap` and cache the state with a TTL (with something like [`ExpiringHashMap`](https://github.com/timberio/vector/blob/2121b7dfa3cdaa342e8a7dbcafb8e3531c36f875/src/expiring_hash_map.rs)). If `evmap` doesn't have the value, the cache will be tried.
Might switch to `Arc`s in the `evmap` values to avoid duplicating `Pod`s in memory.